modules/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/MockHttpProxyServer.java

```diff
@@ -7,91 +7,67 @@
-        serverThread = new Thread(() -> {
-            latch.countDown();
-            while (Thread.currentThread().isInterrupted() == false) {
-                Socket socket;
-                try {
-                    socket = serverSocket.accept();
-                } catch (SocketException e) {
-                    // Server socket is closed
-                    break;
-                } catch (IOException e) {
-                    log.error("Unable to accept socket request", e);
-                    break;
-                }
-                executorService.submit(() -> {
-                    try (
-                        socket;
-                        var is = new BufferedInputStream(socket.getInputStream());
-                        var os = new BufferedOutputStream(socket.getOutputStream())
-                    ) {
-                        // Don't handle keep-alive connections to keep things simple
-                        handler.handle(is, os);
-                    } catch (IOException e) {
-                        log.error("Unable to handle socket request", e);
+    MockHttpProxyServer handler(Supplier<SimpleChannelInboundHandler<FullHttpRequest>> handler) {
+        try {
+            channelFuture = new ServerBootstrap().group(group)
+                .channel(NioServerSocketChannel.class)
+                .childHandler(new ChannelInitializer<SocketChannel>() {
+                    @Override
+                    protected void initChannel(SocketChannel ch) throws Exception {
+                        ch.pipeline()
+                            .addLast(new HttpServerCodec())
+                            .addLast(new HttpObjectAggregator(Integer.MAX_VALUE))
+                            .addLast(handler.get());
                     }
-                });
-            }
-        });
-        serverThread.start();
-    }
-
-    MockHttpProxyServer await() throws InterruptedException {
-        latch.await();
+                })
+                .bind(0)
+                .sync();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException(e);
+        }
+        socketAddress = (InetSocketAddress) channelFuture.channel().localAddress();
         return this;
     }
 
     int getPort() {
-        return serverSocket.getLocalPort();
+        return socketAddress.getPort();
     }
 
     String getHost() {
-        return NetworkAddress.format(serverSocket.getInetAddress());
+        return socketAddress.getHostString();
     }
 
     @Override
     public void close() throws IOException {
-        executorService.shutdown();
-        serverThread.interrupt();
-        serverSocket.close();
-    }
-
-    @FunctionalInterface
-    interface SocketRequestHandler {
-        void handle(InputStream is, OutputStream os) throws IOException;
+        channelFuture.channel().close().awaitUninterruptibly();
+        group.shutdownGracefully();
```

Comment:
You'd want to call `group.shutdownGracefully().awaitUninterruptibly()` here too, otherwise you leak threads and the tests run longer than necessary because the thread leak detector adds a sleep and re-check.

Reply:
Thank you!